[Draft] Resolving integration differences after XGrammar lauch refactoring

gittb opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

After XGrammar released their 0.10 update, there was a good bit of refactoring which seems to have broken the initial integration with SGLang. This PR works to resolve the issues and make XGrammar a working backend for SGLang.

## Modifications

All the changes are going to be within `xgrammar_backend.py`.

So far, we've:
- Fixed the import issues by importing `GrammarCompiler` instead of `CachedGrammarCompiler`
  - Removed the try catch logic that was put in when attempting to handle this
- imported the functions `apply_token_bitmask_inplace` and `allocate_token_bitmask` directly from `matcher`
- removed `vocab_size` parameter from `GrammarMatcher` creation as it's no longer needed by XGrammar
- switched to using `compile_json_schema` rather than `compile_json_schema_grammar` as the method was renamed
- switched to using `clear_cache` instead of `clear` as the method was renamed

I'm chatting with MLC offline in their discord to work through a potential issue between C++ and python within XGrammar

## Checklist

- [ ] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Update documentation as needed, including docstrings or example tutorials.

gittb committed:
Resolving integration differences after xgrammar refactor

merrymercy reviewed (changes requested):
Thanks for the contribution. Can you add a test case similar to this? https://github.com/sgl-project/sglang/blob/main/test/srt/test_json_constrained.py

You can inherit the existing test class and test all existing test cases with xgrammar backend. You can add a class like this in that file.

```python
class TestJSONConstrainedXGrammarBackend(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.model = DEFAULT_MODEL_NAME_FOR_TEST
        cls.base_url = DEFAULT_URL_FOR_TEST
        cls.json_schema = json.dumps(
            {
                "type": "object",
                "properties": {
                    "name": {"type": "string", "pattern": "^[\\w]+$"},
                    "population": {"type": "integer"},
                },
                "required": ["name", "population"],
            }
        )
        cls.process = popen_launch_server(
            cls.model,
            cls.base_url,
            timeout=300,
            other_args=["--max-running-requests", "10", "--grammar-backend", "xgrammar"],
        )
```

Also, can you fix the lint error https://sgl-project.github.io/references/contributor_guide.html?

gittb commented:
Thank you for all your work on SGLang!

Yes, I can do both once I get the implementation working as expected.

Ubospica commented on python/sglang/srt/constrained/xgrammar_backend.py:
```diff
@@ -18,20 +18,18 @@
+    TokenizerInfo,
+)
+
+from xgrammar.matcher import (
```
You can import these functions from `xgrammar` instead of `xgrammar.matcher`

Ubospica commented on python/sglang/srt/constrained/xgrammar_backend.py:
```diff
@@ -86,20 +84,19 @@ def jump_and_retokenize(
         self, vocab_size: int, batch_size: int, device
     ) -> torch.Tensor:
-        return self.matcher.allocate_token_bitmask(vocab_size, batch_size)
+        return allocate_token_bitmask(vocab_size, batch_size)
```
allocate_token_bitmask(batch_size, vocab_size)

gittb replied:
Blunder. Thank you.

Ubospica commented on python/sglang/srt/constrained/xgrammar_backend.py:
```diff
@@ -86,20 +84,19 @@ def jump_and_retokenize(
             self.ctx,
-            max_rollback_tokens=MAX_ROLLBACK_TOKENS,
-            vocab_size=self.vocab_size,
+            max_rollback_tokens=MAX_ROLLBACK_TOKENS
```
You can set `MAX_ROLLBACK_TOKENS` to 200 at line 44. The current 10 is too small.

gittb replied:
Done!

Ubospica commented on python/sglang/srt/constrained/xgrammar_backend.py:
```diff
@@ -112,25 +109,16 @@ def __init__(
-            self.grammar_cache = None
-            return
-
         tokenizer_info = TokenizerInfo.from_huggingface(tokenizer)
```
Assign vocab_size and stop_token_ids (from the chat_template, optionally but will make it more robust) when constructing tokenizer_info

gittb replied:
Agreed.. I had an in issue initially where the vocab_size provided by the tokenizer differed from `vocab_size` param provided when initialing the grammar backend. In my most recent commit I use `vocab_size` when creating `tokenizer_info`. My blocker now is that relating to GrammarMatcher accepting the stop token, but still trying to find the next token mask. I believe this could be related to the backend not having the correct stop tokens.

What is best way to access the chat templates's stop_token_ids information when initializing the grammar backend?

Ubospica commented on python/sglang/srt/constrained/xgrammar_backend.py:
```diff
@@ -112,25 +109,16 @@ def __init__(
-
         tokenizer_info = TokenizerInfo.from_huggingface(tokenizer)
-        self.grammar_cache = CachedGrammarCompiler(tokenizer_info=tokenizer_info)
+        self.grammar_cache = GrammarCompiler(tokenizer_info=tokenizer_info)
```
Would you like to rename grammar_cache to grammar_compiler?

gittb replied:
Yes, updated in the most recent commit

Ubospica reviewed (commented):
Thanks for your contribution! Just a few minor improvements.

gittb committed:
Updated MAX_ROLLBACK_TOKENS
Consolidated imports
Fixed arg order for allocate_token_bitmask
Added logic to move vocab_mask to logits dev prior to apply_token_bitmask_inplace
init xgrammar using vocab_size rather than tokenizer vocab size
rename grammar_cache to grammar_compiler

gittb commented:
@Ubospica Would you take a peek at my logic for moving `vocab_mask` to `logits` device if they are not already on the same device?

My methodology here is that `vocab_mask` must be allocated on the CPU since it must be there to run `fill_next_token_bitmask` but when applying the token bit mask they must be on the same device, I assume logits would be on a cuda device, so simply move the `vocab_mask` to `logits`'s device.

Is this logic correct?

gittb committed:
liting and draft of test case

Ubospica commented:
> My methodology here is that `vocab_mask` must be allocated on the CPU since it must be there to run `fill_next_token_bitmask` but when applying the token bit mask they must be on the same device, I assume logits would be on a cuda device, so simply move the `vocab_mask` to `logits`'s device.

Yes, that is correct. Simply moving it to GPU is okay.

gittb commented:
Closing as work migrated to #2176
